File: packages/python-client/absolute_privacy/mainnet_registry.py
"""Read-only resolution of the checked-in Ethereum mainnet deployment registry."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

DEFAULT_REGISTRY = (
    Path(__file__).resolve().parents[3] / "deployments" / "pools.mainnet.json"
)

MAINNET_ASSET_CHOICES = ("eth", "dai", "lusd")
ASSET_CHOICES = MAINNET_ASSET_CHOICES


def is_mainnet_registry_live(registry: dict[str, Any]) -> bool:
    if registry.get("chainId") != 1:
        return False
    if registry.get("clientsUnlocked") is not True:
        return False
    pools = registry.get("pools") or {}
    for asset_id in ("eth", "dai", "lusd"):
        entry = pools.get(asset_id) or {}
        if not entry.get("pool"):
            return False
    return True


def load_mainnet_registry(path: str | Path | None = None) -> tuple[dict[str, Any], Path]:
    resolved = Path(path).resolve() if path else DEFAULT_REGISTRY.resolve()
    registry = json.loads(resolved.read_text(encoding="utf-8"))
    if registry.get("chainId") != 1 or registry.get("network") != "ethereum-mainnet":
        raise ValueError(f"not a mainnet deployment registry: {resolved}")
    return registry, resolved


def resolve_mainnet_asset(
    asset: str, registry_path: str | Path | None = None
) -> dict[str, Any]:
    registry, resolved = load_mainnet_registry(registry_path)
    asset_id = asset.strip().lower()
    if asset_id in ("weth", "tweth"):
        raise ValueError(
            "unknown mainnet asset 'weth'; native ETH pool is --asset eth"
        )
    entry = (registry.get("pools") or {}).get(asset_id)
    if entry is None:
        choices = ", ".join((registry.get("pools") or {}).keys())
        raise ValueError(f"unknown mainnet asset '{asset}'; choose {choices}")
    if not entry.get("pool"):
        raise ValueError("mainnet pools are not deployed yet. See docs/MAINNET.md")
    return {
        "id": asset_id,
        "chainId": registry["chainId"],
        "network": registry["network"],
        "rpc": registry["rpc"],
        "status": registry["status"],
        "warning": registry["warning"],
        "pool": entry["pool"],
        "token": entry["asset"],
        "symbol": entry["assetSymbol"],
        "decimals": entry["assetDecimals"],
        "source": entry["assetSource"],
        "clientsUnlocked": is_mainnet_registry_live(registry),
        "registryPath": str(resolved),
    }

File: packages/cli/lib/mainnetRegistry.mjs
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { isMainnetRegistryLive } from "../../contracts/scripts/lib/mainnet-safety.mjs";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
export const DEFAULT_MAINNET_REGISTRY = path.resolve(
  __dirname,
  "../../../deployments/pools.mainnet.json"
);

export function loadMainnetRegistry(registryPath = DEFAULT_MAINNET_REGISTRY) {
  const resolved = path.resolve(registryPath);
  const registry = JSON.parse(fs.readFileSync(resolved, "utf8"));
  if (registry.chainId !== 1 || registry.network !== "ethereum-mainnet") {
    throw new Error(`not a mainnet deployment registry: ${resolved}`);
  }
  return { registry, registryPath: resolved };
}

export function resolveMainnetAsset(asset, registryPath = DEFAULT_MAINNET_REGISTRY) {
  const id = String(asset ?? "").trim().toLowerCase();
  if (!id) throw new Error("--asset is required (eth, dai, or lusd)");
  const { registry, registryPath: resolved } = loadMainnetRegistry(registryPath);
  if (id === "weth" || id === "tweth") {
    throw new Error(
      "unknown mainnet asset 'weth'; native ETH pool is --asset eth"
    );
  }
  const entry = registry.pools?.[id];
  if (!entry) {
    const keys = Object.keys(registry.pools ?? {}).join(", ");
    throw new Error(`unknown mainnet asset '${asset}'; choose ${keys}`);
  }
  if (!entry.pool) {
    throw new Error(
      "mainnet pools are not deployed yet. See docs/MAINNET.md"
    );
  }
  return {
    id,
    chainId: registry.chainId,
    network: registry.network,
    rpc: registry.rpc,
    status: registry.status,
    warning: registry.warning,
    pool: entry.pool,
    token: entry.asset,
    symbol: entry.assetSymbol,
    decimals: entry.assetDecimals,
    source: entry.assetSource,
    clientsUnlocked: isMainnetRegistryLive(registry),
    registryPath: resolved,
  };
}

export function resolveNetworkCommandArgs(args, { pool = false, token = false } = {}) {
  if (!args.asset) return args;
  const network = String(args.network ?? "sepolia").toLowerCase();
  if (network === "sepolia") {
    return args;
  }
  if (network !== "mainnet") {
    throw new Error("--network must be sepolia or mainnet");
  }
  const resolved = resolveMainnetAsset(args.asset, args.registry);
  if (pool && !args.pool && !args.to && !args.spender) args.pool = resolved.pool;
  if (token && !args.token) args.token = resolved.token;
  if (args.rpc === undefined || args.rpc === true) args.rpc = resolved.rpc;
  args._resolvedMainnet = resolved;
  return args;
}
